### PR TITLE
Updating result schema to include more specific data

### PR DESCRIPTION
**Pull request by scalvert**
The results that are returned from each failed audit are quite large. When considering the size of the payloads passed through the middleware, passing raw `AxeResults` objects could result in a very large payload. This PR specifically reduces the size of the overall payload that's sent through the middleware, which will also reduce the size written to disc.

In addition to this, there's extra data that's useful to gather when assessing audits, such as
1. routes visited
2. urls visited

The stored result, per test, adheres to this interface:

```ts
interface AxeTestResult {
  moduleName: string;
  testName: string;
  urls: string[];
  routes: string[];
  helpers: string[];
  stack: string;
  violations: Result[];
}
```

<details>
  <summary>JSON result example</summary>

```json
[
  {
    "moduleName": "setupMiddlewareReporter",
    "testName": "gathers results from failed a11yAudit calls",
    "urls": [
      "/"
    ],
    "routes": [
      "violations"
    ],
    "helpers": [
      "visit"
    ],
    "stack": "",
    "violations": [
      {
        "id": "button-name",
        "impact": "critical",
        "tags": [
          "cat.name-role-value",
          "wcag2a",
          "wcag412",
          "section508",
          "section508.22.a"
        ],
        "description": "Ensures buttons have discernible text",
        "help": "Buttons must have discernible text",
        "helpUrl": "https://dequeuniversity.com/rules/axe/4.0/button-name?application=axeAPI",
        "nodes": [
          {
            "any": [
              {
                "id": "button-has-visible-text",
                "data": null,
                "relatedNodes": [],
                "impact": "critical",
                "message": "Element does not have inner text that is visible to screen readers"
              },
              {
                "id": "aria-label",
                "data": null,
                "relatedNodes": [],
                "impact": "serious",
                "message": "aria-label attribute does not exist or is empty"
              },
              {
                "id": "aria-labelledby",
                "data": null,
                "relatedNodes": [],
                "impact": "serious",
                "message": "aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty"
              },
              {
                "id": "role-presentation",
                "data": null,
                "relatedNodes": [],
                "impact": "minor",
                "message": "Element's default semantics were not overridden with role=\"presentation\""
              },
              {
                "id": "role-none",
                "data": null,
                "relatedNodes": [],
                "impact": "minor",
                "message": "Element's default semantics were not overridden with role=\"none\""
              },
              {
                "id": "non-empty-title",
                "data": null,
                "relatedNodes": [],
                "impact": "serious",
                "message": "Element has no title attribute or the title attribute is empty"
              }
            ],
            "all": [],
            "none": [],
            "impact": "critical",
            "html": "<button data-test-selector=\"empty-button\" id=\"violations__empty-button\" class=\"c-button ember-view\">\n</button>",
            "target": [
              "#violations__empty-button"
            ],
            "failureSummary": "Fix any of the following:\n  Element does not have inner text that is visible to screen readers\n  aria-label attribute does not exist or is empty\n  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty\n  Element's default semantics were not overridden with role=\"presentation\"\n  Element's default semantics were not overridden with role=\"none\"\n  Element has no title attribute or the title attribute is empty"
          }
        ]
      },
      {
        "id": "image-alt",
        "impact": "critical",
        "tags": [
          "cat.text-alternatives",
          "wcag2a",
          "wcag111",
          "section508",
          "section508.22.a"
        ],
        "description": "Ensures <img> elements have alternate text or a role of none or presentation",
        "help": "Images must have alternate text",
        "helpUrl": "https://dequeuniversity.com/rules/axe/4.0/image-alt?application=axeAPI",
        "nodes": [
          {
            "any": [
              {
                "id": "has-alt",
                "data": null,
                "relatedNodes": [],
                "impact": "critical",
                "message": "Element does not have an alt attribute"
              },
              {
                "id": "aria-label",
                "data": null,
                "relatedNodes": [],
                "impact": "serious",
                "message": "aria-label attribute does not exist or is empty"
              },
              {
                "id": "aria-labelledby",
                "data": null,
                "relatedNodes": [],
                "impact": "serious",
                "message": "aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty"
              },
              {
                "id": "non-empty-title",
                "data": null,
                "relatedNodes": [],
                "impact": "serious",
                "message": "Element has no title attribute or the title attribute is empty"
              },
              {
                "id": "role-presentation",
                "data": null,
                "relatedNodes": [],
                "impact": "minor",
                "message": "Element's default semantics were not overridden with role=\"presentation\""
              },
              {
                "id": "role-none",
                "data": null,
                "relatedNodes": [],
                "impact": "minor",
                "message": "Element's default semantics were not overridden with role=\"none\""
              }
            ],
            "all": [],
            "none": [],
            "impact": "critical",
            "html": "<img src=\"assets/img/empire-state-building-moonlight.png\" id=\"violations__img-without-alt\" class=\"u-fill c-img ember-view\">",
            "target": [
              "#violations__img-without-alt"
            ],
            "failureSummary": "Fix any of the following:\n  Element does not have an alt attribute\n  aria-label attribute does not exist or is empty\n  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty\n  Element has no title attribute or the title attribute is empty\n  Element's default semantics were not overridden with role=\"presentation\"\n  Element's default semantics were not overridden with role=\"none\""
          }
        ]
      },
      {
        "id": "label",
        "impact": "critical",
        "tags": [
          "cat.forms",
          "wcag2a",
          "wcag412",
          "wcag131",
          "section508",
          "section508.22.n"
        ],
        "description": "Ensures every form element has a label",
        "help": "Form elements must have labels",
        "helpUrl": "https://dequeuniversity.com/rules/axe/4.0/label?application=axeAPI",
        "nodes": [
          {
            "any": [
              {
                "id": "aria-label",
                "data": null,
                "relatedNodes": [],
                "impact": "serious",
                "message": "aria-label attribute does not exist or is empty"
              },
              {
                "id": "aria-labelledby",
                "data": null,
                "relatedNodes": [],
                "impact": "serious",
                "message": "aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty"
              },
              {
                "id": "implicit-label",
                "data": null,
                "relatedNodes": [],
                "impact": "critical",
                "message": "Form element does not have an implicit (wrapped) <label>"
              },
              {
                "id": "explicit-label",
                "data": null,
                "relatedNodes": [],
                "impact": "critical",
                "message": "Form element does not have an explicit <label>"
              },
              {
                "id": "non-empty-title",
                "data": null,
                "relatedNodes": [],
                "impact": "serious",
                "message": "Element has no title attribute or the title attribute is empty"
              }
            ],
            "all": [],
            "none": [],
            "impact": "critical",
            "html": "<input class=\"c-text-input__input\" id=\"violations__labeless-input\" type=\"text\">",
            "target": [
              "#violations__labeless-input"
            ],
            "failureSummary": "Fix any of the following:\n  aria-label attribute does not exist or is empty\n  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty\n  Form element does not have an implicit (wrapped) <label>\n  Form element does not have an explicit <label>\n  Element has no title attribute or the title attribute is empty"
          }
        ]
      }
    ]
  }
]
```

</details>

### CODE REVIEW

File: addon-test-support/setup-middleware-reporter.ts [hunk 1]
```diff
@@ -1,36 +1,70 @@
...
+    currentTestResult = {
+      moduleName: module.name,
+      testName,
+      urls: [currentURL()],
+      routes: [currentRouteName()],
+      helpers: testMetaData.usedHelpers,
+      stack,
+      violations: [],
+    };
+  }
```
Comment (rwjblue) [inline addon-test-support/setup-middleware-reporter.ts:47]:
```suggestion
  if (!currentTestResult) {
    let { module, testName } = QUnit.config.current;
    let testMetaData = getTestMetadata(getContext());

    let stack = (!DEBUG && new Error().stack) || '';

    currentTestResult = {
      moduleName: module.name,
      testName,
      urls: new Set(),
      routes: new Set(),
      helpers: testMetaData.usedHelpers,
      stack,
      violations: [],
    };
  }

  currentTestResult.urls.add(currentURL());
  currentTestResult.routes.add(currentRouteName());
```

File: addon-test-support/setup-middleware-reporter.ts [hunk 2]
```diff
@@ -1,36 +1,70 @@
...
+    }
+  });
 }
 
-export async function middlewareReporter(axeResults: AxeResults) {
-  TEST_SUITE_RESULTS.push(buildResult(axeResults));
+export function pushTestResult() {
+  if (typeof currentTestResult !== 'undefined') {
+    currentTestResult.violations = [...currentViolationsMap.values()];
+    TEST_SUITE_RESULTS.push(currentTestResult);
```
Comment (rwjblue) [inline addon-test-support/setup-middleware-reporter.ts:63]:
```suggestion
    currentTestResults.urls = [...currentTestResults.urls.values()]);
    currentTestResults.routes = [...currentTestResults.routes.values()]);
    TEST_SUITE_RESULTS.push(currentTestResult);
```

File: addon-test-support/setup-middleware-reporter.ts [hunk 3]
```diff
@@ -1,36 +1,70 @@
...
+
+    currentTestResult = undefined;
+    currentViolationsMap = new Map();
```
Comment (rwjblue) [inline addon-test-support/setup-middleware-reporter.ts:66]: It's vaguely odd to reset one to `new Map()` and the other to `undefined`. I guess I'd prefer to set both back to `undefined` here, then initialize inside the `if (!currentTestResult) {` branch above.

File: setup-middleware.js [hunk 4]
```diff
@@ -13,6 +13,7 @@ function reportViolations(req, res, options) {
   );
 
   ensureDirSync(outputDir);
+  emptyDirSync(outputDir);
```
Comment (rwjblue) [inline setup-middleware.js:16]: I don't think this is quite right, you can't empty the dir before writing each file (you'd only ever end up with one file in a directory).

I think we need to `ensureDirSync(); emptyDirSync()` down in the `setupMiddleware` function (so it happens once when we get started, not once per POST).

File: addon-test-support/setup-middleware-reporter.ts (hunk 1 above)
Comment (scalvert) [reply addon-test-support/setup-middleware-reporter.ts:47]: Are you gonna mess with my formatting again, @rwjblue??? :D

File: addon-test-support/setup-middleware-reporter.ts (hunk 3 above)
Comment (scalvert) [reply addon-test-support/setup-middleware-reporter.ts:71]: 🤷

File: setup-middleware.js (hunk 4 above)
Comment (scalvert) [reply setup-middleware.js:16]: Ah yes, you're right. I can fix this up.

**Comment by MelSumner** [comment]
> The results that are returned from each failed audit are quite large.

@scalvert Would you please post a sample result, so we can compare/contrast what will be audited out with this change?

File: addon-test-support/setup-middleware-reporter.ts (hunk 3 above)
Comment (scalvert) [reply addon-test-support/setup-middleware-reporter.ts]: I changed it to make it more consistent.

File: setup-middleware.js (hunk 4 above)
Comment (scalvert) [reply setup-middleware.js]: Fixed. This was a good catch. There's a lot of moving parts here and I need to figure out a way to make it more understandable.

**Comment by scalvert** [comment]
@MelSumner sure thing. I've added a sample output. The TypeScript interface effectively gives you a sense of the format of the output.

**Comment by scalvert** [comment]
@rwjblue @MelSumner any more updates/feedback/issues with this PR? I'd like to get it shipped soon so we can target it for a new release.

**Review by drewlee — APPROVED** [review]

**Pull request merged**